src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -337,87 +354,279 @@ out ImmutableArray<(BoundExpression, BoundDagTemp)> bindings)
+                {
+                    case BoundDagEvaluation e:
+                        var tail = MakeDecisionDag(syntax, RemoveEvaluation(cases, e), defaultLabel);
+                        return new BoundEvaluationPoint(syntax, e, tail);
+                    case BoundDagDecision d:
+                        SplitCases(cases, d, out ImmutableArray<PartialCaseDecision> whenTrueDecisions, out ImmutableArray<PartialCaseDecision> whenFalseDecisions);
+                        var whenTrue = MakeDecisionDag(syntax, whenTrueDecisions, defaultLabel);
+                        var whenFalse = MakeDecisionDag(syntax, whenFalseDecisions, defaultLabel);
+                        return new BoundDecisionPoint(syntax, d, whenTrue, whenFalse);
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(first.Decisions[0].Kind);
+                }
+            }
         }
-    }
 
-    partial class BoundDagEvaluation
-    {
-        public override bool Equals(object obj) => obj is BoundDagEvaluation other && this.Equals(other);
-        public bool Equals(BoundDagEvaluation other)
+        private void SplitCases(
+            ImmutableArray<PartialCaseDecision> cases,
+            BoundDagDecision d,
+            out ImmutableArray<PartialCaseDecision> whenTrue,
+            out ImmutableArray<PartialCaseDecision> whenFalse)
         {
-            return other != (object)null && this.Kind == other.Kind && this.Input.Equals(other.Input) && this.Symbol == other.Symbol;
+            var whenTrueBuilder = ArrayBuilder<PartialCaseDecision>.GetInstance();
+            var whenFalseBuilder = ArrayBuilder<PartialCaseDecision>.GetInstance();
+            foreach (var c in cases)
+            {
+                FilterCase(c, d, whenTrueBuilder, whenFalseBuilder);
+            }
+
+            whenTrue = whenTrueBuilder.ToImmutableAndFree();
+            whenFalse = whenFalseBuilder.ToImmutableAndFree();
         }
-        private Symbol Symbol
+
+        private void FilterCase(
+            PartialCaseDecision c,
+            BoundDagDecision d,
+            ArrayBuilder<PartialCaseDecision> whenTrueBuilder,
+            ArrayBuilder<PartialCaseDecision> whenFalseBuilder)
         {
-            get
+            var trueBuilder = ArrayBuilder<BoundDagDecision>.GetInstance();
+            var falseBuilder = ArrayBuilder<BoundDagDecision>.GetInstance();
+            foreach (var dd in c.Decisions)
             {
-                switch (this)
+                CheckConsistentDecision(d: d, other: dd,
+                    permitsTrue: out bool permitsTrue, permitsFalse: out bool permitsFalse,
+                    killsDecisionOnTrueBranch: out bool killsDecisionOnTrueBranch,
+                    killsDecisionOnFalseBranch: out bool killsDecisionOnFalseBranch);
+                if (permitsTrue)
+                {
+                    if (d != dd && !killsDecisionOnTrueBranch) trueBuilder?.Add(dd);
```

Comment:
>trueBuilder?.Add(dd) [](start = 63, length = 20)

Please put ```{}``` around embedded statement.

Reply:
Addressed in https://github.com/dotnet/roslyn/pull/24059/commits/114a56e9077ed03f08443889b69cae6972a4acc6 of https://github.com/dotnet/roslyn/pull/24059